module/os/linux/zfs/vdev_disk.c

```diff
@@ -670,126 +591,488 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+		*pagep = page;
+		*offp = offset_in_page(buf);
+		return;
+	}
 
 	/*
-	 * Since bio's can have up to BIO_MAX_PAGES=256 iovec's, each of which
-	 * is at least 512 bytes and at most PAGESIZE (typically 4K), one bio
-	 * can cover at least 128KB and at most 1MB.  When the required number
-	 * of iovec's exceeds this, we are forced to break the IO in multiple
-	 * bio's and wait for them all to complete.  This is likely if the
-	 * recordsize property is increased beyond 1MB.  The default
-	 * bio_count=16 should typically accommodate the maximum-size zio of
-	 * 16MB.
+	 * This page is part of a "compound page", which is a group of pages
+	 * that can be referenced from a single struct page *. Its organised as
+	 * a "head" page, followed by a serious of "tail" pages.
+	 *
+	 * In OpenZFS, compound pages are allocated using the __GFP_COMP flag,
+	 * which we get from scatter ABDs and SPL vmalloc slabs (ie >16K
+	 * allocations). So a great many of the IO buffers we get are going to
+	 * be of this type.
+	 *
+	 * The tail pages are just regular PAGE_SIZE pages, and we can just
+	 * load them into the BIO the same as we would for non-compound pages
+	 * above, and it all works just fine. However, the head page has length
+	 * covering itself and all the tail pages. If our buffer spans multiple
+	 * pages, then we can load the head page and a >PAGE_SIZE length into
+	 * the BIO, which is far more efficient.
+	 *
+	 * To do this, we need to calculate the offset of the buffer from the
+	 * head page (offset_in_page() is the offset within its PAGE_SIZE'd
+	 * page ie just a simple ~(PAGE_SIZE-1) mask).
 	 */
 
-	abd_offset = 0;
-	bio_offset = io_offset;
-	bio_size = io_size;
-	for (int i = 0; i <= dr->dr_bio_count; i++) {
+	*pagep = compound_head(page);
+	*offp = (uint_t)((uintptr_t)(buf) & (page_size(*pagep)-1));
+}
 
-		/* Finished constructing bio's for given buffer */
-		if (bio_size <= 0)
-			break;
+/*
+ * Iterator callback to count ABD pages and check their size & alignment.
+ *
+ * On Linux, each BIO segment can take a page pointer, and an offset+length of
+ * the data within that page. A page can be arbitrarily large ("compound"
+ * pages) but we still have to ensure the data portion is correctly sized and
+ * aligned to the logical block size, to ensure that if the kernel wants to
+ * split the BIO, the two halves will still be properly aligned.
+ */
+typedef struct {
+	uint_t	bmask;
+	uint_t	npages;
+	uint_t	end;
+} vdev_disk_check_pages_t;
+
+static int
+vdev_disk_check_pages_cb(void *buf, size_t len, void *priv)
+{
+	vdev_disk_check_pages_t *s = priv;
+
+	struct page *page;
+	uint_t off;
+
+	/*
+	 * If we didn't finish on a block size boundary last time, then there
+	 * would be a gap if we tried to use this ABD as-is, so abort.
+	 */
+	if (s->end != 0)
+		return (1);
+
+	/*
+	 * Note if we're taking less than a full block, so we can check it
+	 * above on the next call.
+	 */
+	s->end = len & s->bmask;
```

Comment:
I could just leave it at PAGE_SIZE, but I found with a 512b LBS and pushing everything into tiny allocations (`metaslab_force_ganging=32768`, `metaslab_force_ganging_pct=100`) it was constantly finding gaps it couldn't resolve, and had to linearise. Granted, that's an extreme case, but it seemed no big deal to just use the block size.

Ironically, the kernel should basically never need to split a BIO anymore, because we're never overfilling it. I suppose it gets trickier once various DM layers are in place, as we saw with LUKS, but I doubt its gonna be that much of an issue.

> What Linux will do...

Something bonkers, I'm sure.